### PR TITLE
Include classname of null-returning `map` function in NPE msg

### PR DESCRIPTION
**Pull request by ismailalammar**
linked issue #2982 

this PR include  : 

- log class name in (FluxMap - FluxMapSignal - FluxMapFuseable) if mapper returns null.

### CODE REVIEW

**Comment by simonbasle** [comment]
thanks @ismailalammar !
as I stated in the issue, this should have been opened against the `3.4.x` branch.
you will need to `git rebase -i 3.4.x` and `drop` the commits that don't belong to the PR, then I can edit the PR target (or you can do it if you're familiar with the procedure).

Also, can you try to add tests for these in `FluxMapTest` and `FluxMapSignalTest`?
In `FluxMapTest` there is a `mapperReturnsNull` test but it is using an old style (no `StepVerifier`), doesn't assert the message of the `NullPointerException` and only covers a fused case.

File: reactor-core/src/main/java/reactor/core/publisher/FluxMap.java [hunk 1]
```diff
@@ -104,7 +104,7 @@ public void onNext(T t) {
 
 			try {
 				v = Objects.requireNonNull(mapper.apply(t),
-						"The mapper returned a null value.");
+						"The mapper [" + mapper.getClass().getName() + "] returned a null value.");
```
Comment (JamesChenX) [inline reactor-core/src/main/java/reactor/core/publisher/FluxMap.java:107]: For better performance, we can concat the strings only after checking the return value is null. e.g.:
```java
Object value = mapper.apply(t);
if (value == null) {
    throw new NullPointerException("The mapper [" + mapper.getClass().getName() + "] returned a null value.");
}
```

Comment (ismailalammar) [reply reactor-core/src/main/java/reactor/core/publisher/FluxMap.java:107]: sure will change it , thanks @JamesChenX

**Comment by ismailalammar** [comment]
> thanks @ismailalammar ! as I stated in the issue, this should have been opened against the `3.4.x` branch. you will need to `git rebase -i 3.4.x` and `drop` the commits that don't belong to the PR, then I can edit the PR target (or you can do it if you're familiar with the procedure).
> 
> Also, can you try to add tests for these in `FluxMapTest` and `FluxMapSignalTest`? In `FluxMapTest` there is a `mapperReturnsNull` test but it is using an old style (no `StepVerifier`), doesn't assert the message of the `NullPointerException` and only covers a fused case.

sure , will do

**Comment by simonbasle** [comment]
@ismailalammar any progress on this? do you need help with writing the tests?

**Comment by hepin1989** [comment]
Can it rightly retrieve the lambda's line numbers too?

**Comment by simonbasle** [comment]
> Can it rightly retrieve the lambda's line numbers too?

I would rather rely on Java 14+ "better NPE" feature

**Comment by ismailalammar** [comment]
i am so sorry i missed this ticket , got blocked with amount of work in the past few weeks. 

i don't think i will be able to complete the test cases in the near future . apologize for that

**Review by simonbasle — APPROVED** [review]

**Pull request merged**